File: packages/core/src/propagation.ts
import type { EventType, PulseEvent, Rule } from './types.js'
import { Mailbox } from './mailbox.js'
import { DAG } from './dag.js'

/**
 * Core propagation loop.
 * Drains mailboxes in topological (DAG) order, fires rules.
 * Repeats until quiescent or max rounds exceeded.
 *
 * For 'each' mode rules: fires the handler for each pending event in each trigger's mailbox.
 * For 'join' mode rules: fires only when all trigger mailboxes have at least one ready event.
 *
 * Handlers that emit events do so via the engine's emit path (which deposits into
 * mailboxes and traces DAG edges). The propagation loop picks up newly deposited
 * events on the next iteration.
 */
export function propagate(
  dag: DAG,
  mailboxes: Map<EventType, Mailbox>,
  maxRounds: number,
  onError?: (error: Error, rule: Rule, event: any) => void,
): void {
  // Evict orphaned events whose consumers have all been disposed
  for (const mb of mailboxes.values()) {
    mb.evictOrphans()
  }

  let rounds = 0
  let work = true

  while (work) {
    if (rounds >= maxRounds) {
      throw new Error(`Propagation exceeded ${maxRounds} rounds — probable cycle in rule graph`)
    }
    work = false
    rounds++

    const order = dag.getTopologicalOrder()
    for (const rule of order) {
      if (rule._disposed) continue

      if (rule.mode === 'each') {
        // Each-mode: fire for every trigger that has a ready event
        for (const trigger of rule.triggers) {
          const mb = getMailbox(mailboxes, trigger)
          while (mb.hasReadyEvent(rule)) {
            const ev = mb.consume(rule)
            if (rule.guard && !rule.guard(ev.payload)) continue
            try {
              rule.action(ev.payload)
              work = true
            } catch (err) {
              if (onError) {
                onError(err as Error, rule, ev.payload)
              } else {
                throw err
              }
            }
          }
        }
      } else if (rule.mode === 'join') {
        let allReady = true
        for (const t of rule.triggers) {
          if (!getMailbox(mailboxes, t).hasReadyEvent(rule)) {
            allReady = false
            break
          }
        }
        if (allReady) {
          const consumed: PulseEvent[] = []
          const payloads: any[] = []
          for (const t of rule.triggers) {
            const ev = getMailbox(mailboxes, t).consume(rule)
            consumed.push(ev)
            payloads.push(ev.payload)
          }
          if (rule.guard && !rule.guard(...payloads)) {
            for (let i = 0; i < consumed.length; i++) {
              getMailbox(mailboxes, rule.triggers[i]).unconsumeAll(rule, [consumed[i]])
            }
            continue
          }
          try {
            rule.action(...payloads)
            work = true
          } catch (err) {
            if (onError) {
              onError(err as Error, rule, payloads)
            } else {
              throw err
            }
          }
        }
      }
    }
  }
}

function getMailbox<T>(mailboxes: Map<EventType, Mailbox>, type: EventType<T>): Mailbox<T> {
  let mb = mailboxes.get(type)
  if (!mb) {
    mb = new Mailbox(type)
    mailboxes.set(type, mb)
  }
  return mb as Mailbox<T>
}

File: packages/dart/lib/src/propagation.dart
import 'event_type.dart';
import 'event.dart';
import 'mailbox.dart';
import 'rule.dart';
import 'dag.dart';
import 'skip.dart';

/// Core propagation loop.
/// Drains mailboxes in topological (DAG) order, fires rules, deposits resulting events.
/// Repeats until quiescent or max rounds exceeded.
void propagate({
  required DAG dag,
  required Map<EventType, Mailbox> mailboxes,
  required void Function(EventType type, dynamic payload) emitFn,
  required int maxRounds,
  void Function(Object error, Rule rule, dynamic event)? onError,
}) {
  var rounds = 0;
  var work = true;

  while (work) {
    if (rounds >= maxRounds) {
      throw StateError(
          'Propagation exceeded $maxRounds rounds \u2014 probable cycle in rule graph');
    }
    work = false;
    rounds++;

    final order = dag.getTopologicalOrder();
    for (final rule in order) {
      if (rule.disposed) continue;

      if (rule.mode == RuleMode.each) {
        final trigger = rule.triggers[0];
        final mb = _getMailbox(mailboxes, trigger);
        while (mb.hasReadyEvent(rule)) {
          final ev = mb.consume(rule);
          if (rule.guard != null && !rule.guard!([ev.payload])) continue;
          try {
            final result = rule.action([ev.payload]);
            if (result != null && !isSkip(result) && rule.outputs.isNotEmpty) {
              if (rule.outputs.length > 1 && result is List) {
                for (var i = 0; i < rule.outputs.length; i++) {
                  _depositEvent(mailboxes, rule.outputs[i], result[i]);
                }
              } else {
                _depositEvent(mailboxes, rule.outputs[0], result);
              }
              work = true;
            }
          } catch (err) {
            if (onError != null) {
              onError(err, rule, ev.payload);
            } else {
              rethrow;
            }
          }
        }
      } else if (rule.mode == RuleMode.join) {
        var allReady = true;
        for (final t in rule.triggers) {
          if (!_getMailbox(mailboxes, t).hasReadyEvent(rule)) {
            allReady = false;
            break;
          }
        }
        if (allReady) {
          final consumed = <PulseEvent>[];
          final payloads = <dynamic>[];
          for (final t in rule.triggers) {
            final ev = _getMailbox(mailboxes, t).consume(rule);
            consumed.add(ev);
            payloads.add(ev.payload);
          }
          if (rule.guard != null && !rule.guard!(payloads)) {
            for (var i = 0; i < consumed.length; i++) {
              _getMailbox(mailboxes, rule.triggers[i])
                  .unconsumeAll(rule, [consumed[i]]);
            }
            continue;
          }
          try {
            final result = rule.action(payloads);
            if (result != null && !isSkip(result) && rule.outputs.isNotEmpty) {
              _depositEvent(mailboxes, rule.outputs[0], result);
              work = true;
            }
          } catch (err) {
            if (onError != null) {
              onError(err, rule, payloads);
            } else {
              rethrow;
            }
          }
        }
      }
    }
  }
}

Mailbox _getMailbox(Map<EventType, Mailbox> mailboxes, EventType type) {
  return mailboxes.putIfAbsent(type, () => Mailbox(type));
}

void _depositEvent(
    Map<EventType, Mailbox> mailboxes, EventType type, dynamic payload) {
  final ev = createEvent(type, payload);
  if (ev.pendingConsumers.isNotEmpty) {
    _getMailbox(mailboxes, type).enqueue(ev);
  }
}
